### PR TITLE
Refactor/classification 10

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Continues work on classification refactor https://github.com/Lightning-AI/metrics/issues/1001
Prior work:

* https://github.com/Lightning-AI/metrics/pull/1189
* https://github.com/Lightning-AI/metrics/pull/1175
* https://github.com/Lightning-AI/metrics/pull/1167
* https://github.com/Lightning-AI/metrics/pull/1163
* https://github.com/Lightning-AI/metrics/pull/1159
* https://github.com/Lightning-AI/metrics/pull/1151
* https://github.com/Lightning-AI/metrics/pull/1145
* https://github.com/Lightning-AI/metrics/pull/1143
* https://github.com/Lightning-AI/metrics/pull/1054

This PR is introducing the changes for the transition (deprecation) of the old API to the new.

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — COMMENTED** [review]
Shall we also remove all the commented-out code like tests?

**Comment by SkafteNicki** [comment]
> Shall we also remove all the commented-out code like tests?

Yes.
There may be some of the tests that we want to reintroduce but we should still be able to find them in the future.

**Comment by Borda** [comment]
@SkafteNicki still draft? :innocent:

**Comment by SkafteNicki** [comment]
@Borda ready for review

File: src/torchmetrics/classification/average_precision.py [hunk 1]
```diff
@@ -384,11 +384,45 @@ class AveragePrecision(Metric):
...
+                )
+            elif task == "multilabel":
+                return MultilabelAveragePrecision(
+                    num_labels, average, thresholds, ignore_index, validate_args, **kwargs
+                )
+            else:
+                raise ValueError(
+                    f"Expected argument `task` to either be `'binary'`, `'multiclass'` or `'multilabel'` but got {task}"
+                )
+        return super().__new__(cls)
```
Comment (Borda) [inline src/torchmetrics/classification/average_precision.py:414]:
```suggestion
            if task == "multiclass":
                return MulticlassAveragePrecision(
                    num_classes, average, thresholds, ignore_index, validate_args, **kwargs
                )
            if task == "multilabel":
                return MultilabelAveragePrecision(
                    num_labels, average, thresholds, ignore_index, validate_args, **kwargs
                )
            raise ValueError(
                f"Expected argument `task` to either be `'binary'`, `'multiclass'` or `'multilabel'` but got {task}"
            )
        return super().__new__(cls)
```

File: src/torchmetrics/classification/auroc.py [hunk 2]
```diff
@@ -404,12 +404,43 @@ class AUROC(Metric):
...
+            if task == "binary":
+                return BinaryAUROC(max_fpr, thresholds, ignore_index, validate_args, **kwargs)
+            elif task == "multiclass":
+                return MulticlassAUROC(num_classes, average, thresholds, ignore_index, validate_args, **kwargs)
+            elif task == "multilabel":
+                return MultilabelAUROC(num_labels, average, thresholds, ignore_index, validate_args, **kwargs)
+            else:
+                raise ValueError(
+                    f"Expected argument `task` to either be `'binary'`, `'multiclass'` or `'multilabel'` but got {task}"
+                )
```
Comment (Borda) [inline src/torchmetrics/classification/auroc.py:430]:
```suggestion
            if task == "multiclass":
                return MulticlassAUROC(num_classes, average, thresholds, ignore_index, validate_args, **kwargs)
            if task == "multilabel":
                return MultilabelAUROC(num_labels, average, thresholds, ignore_index, validate_args, **kwargs)
            raise ValueError(
                f"Expected argument `task` to either be `'binary'`, `'multiclass'` or `'multilabel'` but got {task}"
            )
```

File: src/torchmetrics/classification/binned_precision_recall.py [hunk 3]
```diff
@@ -283,6 +314,11 @@ def __init__(
         thresholds: Union[int, Tensor, List[float]] = 100,
         **kwargs: Any,
     ) -> None:
+        rank_zero_warn(
+            "Metric `BinnedRecallAtFixedPrecision` has been deprecated in v0.10 and will be completly removed in v0.11."
+            "Instead use the refactored version of `RecallAtFixedPrecision` by specifying the `thresholds` argument.",
```
Comment (Borda) [inline src/torchmetrics/classification/binned_precision_recall.py:319]:
```suggestion
            " Instead, use the refactored version of `RecallAtFixedPrecision` by specifying the `thresholds` argument.",
```

File: src/torchmetrics/classification/binned_precision_recall.py [hunk 4]
```diff
@@ -225,6 +239,19 @@ class BinnedAveragePrecision(BinnedPrecisionRecallCurve):
...
 
+    def __init__(
+        self,
+        num_classes: int,
+        thresholds: Union[int, Tensor, List[float]] = 100,
+        **kwargs: Any,
+    ) -> None:
+        rank_zero_warn(
+            "Metric `BinnedAveragePrecision` has been deprecated in v0.10 and will be completly removed in v0.11."
+            "Instead use the refactored version of `AveragePrecision` by specifying the `thresholds` argument.",
```
Comment (Borda) [inline src/torchmetrics/classification/binned_precision_recall.py:250]:
```suggestion
            " Instead, use the refactored version of `AveragePrecision` by specifying the `thresholds` argument.",
```

File: src/torchmetrics/classification/binned_precision_recall.py [hunk 5]
```diff
@@ -122,6 +127,11 @@ def __init__(
         thresholds: Union[int, Tensor, List[float]] = 100,
         **kwargs: Any,
     ) -> None:
+        rank_zero_warn(
+            "Metric `BinnedPrecisionRecallCurve` has been deprecated in v0.10 and will be completly removed in v0.11."
+            "Instead use the refactored version of `PrecisionRecallCurve` by specifying the `thresholds` argument.",
```
Comment (Borda) [inline src/torchmetrics/classification/binned_precision_recall.py:132]:
```suggestion
            " Instead use the refactored version of `PrecisionRecallCurve` by specifying the `thresholds` argument.",
```

File: src/torchmetrics/classification/calibration_error.py [hunk 6]
```diff
@@ -265,6 +265,27 @@ class CalibrationError(Metric):
...
+    ) -> None:
+        if task is not None:
+            if task == "binary":
+                return BinaryCalibrationError(n_bins, norm, ignore_index, validate_args, **kwargs)
+            elif task == "multiclass":
+                return MulticlassCalibrationError(num_classes, n_bins, norm, ignore_index, validate_args, **kwargs)
+            else:
+                raise ValueError(
+                    f"Expected argument `task` to either be `'binary'`, `'multiclass'` or `'multilabel'` but got {task}"
+                )
```
Comment (Borda) [inline src/torchmetrics/classification/calibration_error.py:286]:
```suggestion
            if task == "binary":
                return BinaryCalibrationError(n_bins, norm, ignore_index, validate_args, **kwargs)
            if task == "multiclass":
                return MulticlassCalibrationError(num_classes, n_bins, norm, ignore_index, validate_args, **kwargs)
            raise ValueError(
                f"Expected argument `task` to either be `'binary'`, `'multiclass'` or `'multilabel'` but got {task}"
            )
```

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
Could you pls check the typing?

**Comment by SkafteNicki** [comment]
@Borda not sure what to do about the typing. Take `jaccard_score` as an example:
```python
def jaccard_index(
    preds: Tensor,
    target: Tensor,
    num_classes: int,
    average: Optional[str] = "macro",
    ignore_index: Optional[int] = None,
    absent_score: float = 0.0,
    threshold: float = 0.5,
    task: Optional[Literal["binary", "multiclass", "multilabel"]] = None,  # new arg
    num_labels: Optional[int] = None,  # new arg
    validate_args: bool = True,  # new arg
) -> Tensor:
    if task is not None:
        kwargs = dict(ignore_index=ignore_index, validate_args=validate_args)
        if task == "binary":
            return binary_jaccard_index(preds, target, threshold, **kwargs)
        if task == "multiclass":
            return multiclass_jaccard_index(preds, target, num_classes, average, **kwargs)
        if task == "multilabel":
            return multilabel_jaccard_index(preds, target, num_labels, threshold, average, **kwargs)
        raise ValueError(
            f"Expected argument `task` to either be `'binary'`, `'multiclass'` or `'multilabel'` but got {task}"
        )
    else:
        rank_zero_warn(
            ...
            DeprecationWarning,
        )
    confmat = _confusion_matrix_update(preds, target, num_classes, threshold)
    return _jaccard_from_confmat(confmat, num_classes, average, ignore_index, absent_score)
```
The new argument `num_labels` is an optional argument because it is not necessary for the old implementation or in the case of `task='binary'` and `task='multiclass'`, but it is an required parameter if user set `task='multilabel'`. Implementation wise should it be pretty safe to call it like this because all arguments are validated inside the sub-functions however it still leads to a type-mismatch.

**Review by justusschock — APPROVED** [review]

**Pull request merged**